### PR TITLE
while set parallel to classes the program may occur the exception:

### PR DESCRIPTION
**Pull request by jianhua3331**
Set `parallel = classes` and `thread-count = 20`.
If there are a lot of testcase failed or throw skiptest, the program may occur the exception(Lots):
```
java.util.ConcurrentModificationException
at java.util.Hashtable.next(Hashtable.java:1167)
at org.testng.internal.Invoker.classConfigurationFailed(Invoker.java:401)
...
```

Fixes # .

### Did you remember to?

- [ ] Add test case(s)
- [x] Update `CHANGES.txt`

### CODE REVIEW

File: src/main/java/org/testng/internal/Invoker.java [hunk 1]
```diff
@@ -88,12 +88,14 @@
   private static final Predicate<ITestNGMethod, IClass> SAME_CLASS = new SameClassNamePredicate();
 
   private void setClassInvocationFailure(Class<?> clazz, Object instance) {
-    Set<Object> instances = m_classInvocationResults.get( clazz );
-    if (instances == null) {
-      instances = Sets.newHashSet();
-      m_classInvocationResults.put(clazz, instances);
+    synchronized(this){
```
Comment (juherr) [inline src/main/java/org/testng/internal/Invoker.java]: As I know, `synchronized(this)` is not a good practice, and in your case, you just want a synchronization on `m_classInvocationResults`.

**Review by juherr — CHANGES_REQUESTED** [review]
I suppose the issue is too random to be able to have a non-regression test but `CHANGES.txt` must be updated.

**Comment by juherr** [comment]
@jianhua3331 Why did you close the pull request?

**Comment by jianhua3331** [comment]
Sorry, I have not seen the message before close it.
I mistakenly think the submission was rejected.
Is it Only need to add the CHANGES.txt in this pull request?

**Comment by juherr** [comment]
Yep, and I'd like to be sure about `synchronized(this)` too. I'm not feeling it good for the moment.

If you are able to make a test, it will be perfect but as I said, I'm not sure it is possible.

BTW, few effort before the merge IMO ;)

**Comment by jianhua3331** [comment]
I have test it in our project and the problem is resolved.
We have a recurring scene which will throw much of SkipTest exceptions.

**Comment by juherr** [comment]
> I have test it in our project and the problem is resolved.

In fact, the idea behind the test is: it must fail before the fix, it must pass after the fix and it will guarantee the detection of the regression in the future.
Could you add one?

Thanks for Change.txt. What about `synchronized(this)`?

File: src/main/java/org/testng/internal/Invoker.java (hunk 1 above)
Comment (jianhua3331) [inline src/main/java/org/testng/internal/Invoker.java]: Yes, you are right, I changed to synchronization on m_classInvocationResults.

File: CHANGES.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿Current
```
Comment (juherr) [inline CHANGES.txt]: Please, keep this line.
And you can add your name to the list. You made the job, you deserve it! :)

**Review by juherr — APPROVED** [review]

**Comment by juherr** [comment]
LGTM 👍 Ping @cbeust

File: CHANGES.txt [hunk 3]
```diff
@@ -1,4 +1,5 @@
...
+Current
+Fixed: GITHUB-1302: while set parallel to classes the program may occur the exception(Jianhua Li)
```
Comment (cbeust) [inline CHANGES.txt]: Rephrase: "When 'parallel' is set to 'classes', `ConcurrentModificationException` can be thrown.

**Pull request merged**